ProcessLib/BalanceProcess/BalanceProcessLocalAssembler.h

```diff
@@ -0,0 +1,154 @@
+    using NodalVectorType = typename ShapeMatricesType::NodalVectorType;
+
+public:
+    /// Precomputes the shape matrices for a given surface element.
+    /// @param surface_element The surface element used for precomputing the
+    /// shape matrices used later on for the integration.
+    /// @param integration_order the order of the integration
+    /// @param mesh Needed for fetching the id of the bulk element and the id of
+    /// the face
+    BalanceProcessLocalAssembler(
+        MeshLib::Element const& surface_element,
+        std::size_t const /*local_matrix_size*/,
+        unsigned const integration_order,
+        MeshLib::PropertyVector<std::size_t> const& bulk_element_ids,
+        MeshLib::PropertyVector<std::size_t> const& bulk_face_ids)
+        : _surface_element(surface_element),
+          _integration_method(integration_order),
+          _bulk_element_id(bulk_element_ids[surface_element.getID()]),
+          _bulk_face_id(bulk_face_ids[surface_element.getID()])
+    {
+        // only detJ of surface element is needed
+        using FemType =
+            NumLib::TemplateIsoparametric<ShapeFunction, ShapeMatricesType>;
+
+        FemType fe(*static_cast<const typename ShapeFunction::MeshElement*>(
+            &surface_element));
+
+        std::size_t const n_integration_points =
+            _integration_method.getNumberOfPoints();
+
+        std::vector<typename ShapeMatricesType::ShapeMatrices> shape_matrices;
+        shape_matrices.reserve(n_integration_points);
+        _detJ.reserve(n_integration_points);
+        for (std::size_t ip = 0; ip < n_integration_points; ++ip)
+        {
+            shape_matrices.emplace_back(ShapeFunction::DIM, GlobalDim,
+                                        ShapeFunction::NPOINTS);
+            // TODO computeShapeFunctions only detJ necessary
+            fe.computeShapeFunctions(
+                _integration_method.getWeightedPoint(ip).getCoords(),
+                shape_matrices[ip], GlobalDim);
+            _detJ.push_back(shape_matrices[ip].detJ);
+        }
+    }
+
+    /// Integration for the element with the id \c element_id.
+    /// @param element_id id of the element
+    /// @param x The global vector containing the values for numerical
+    /// integration.
+    /// @param balance PropertyVector the integration result will be stored
+    /// into, where balance has the same number of entries like mesh elements
+    /// exists.
+    /// @param bulk_process Reference to the bulk process is used to compute the
+    /// flux.
+    void integrate(std::size_t element_id,
+                   GlobalVector const& x,
+                   MeshLib::PropertyVector<double>& balance,
+                   Process const& bulk_process) override
+    {
+        SpatialPosition pos;
+        pos.setElementID(element_id);
+
+        auto surface_element_normal =
+            MeshLib::FaceRule::getSurfaceNormal(&_surface_element);
```

Comment:
> The higher order elements are geometrically linear.

AFAIK that is not true in general. The higher-order nodes are not necessarily on the faces, but can deviate from them. I think, @nagelt can explain that better than I.